CRM/Core/Permission/Joomla.php

```diff
@@ -131,4 +135,74 @@ public function checkGroupRole($array) {
     return FALSE;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function upgradePermissions($permissions) {
+    $translatedPerms = array();
+    foreach (array_flip($permissions) as $perm) {
+      $translated = $this->translateJoomlaPermission($perm);
+      $translatedPerms[] = $translated[0];
+    }
+
+    $associations = $this->getUserGroupPermsAssociations();
+    $originalAssociations = clone $associations;
+    foreach (array_keys(get_object_vars($associations)) as $permName) {
+      if (!in_array($permName, $translatedPerms)) {
+        unset($associations->$permName);
+      }
+    }
+
+    if ($originalAssociations !== $associations) {
```

Comment:
referencing this: https://secure.php.net/manual/en/language.oop5.object-comparison.php

...I think this will always be true. I think you need to set a flag in the loop above if you have perms to update.